org.elysium.parent/org.elysium.ui/src/org/elysium/ui/contentassist/LilyPondProposalProvider.java

```diff
@@ -41,33 +74,114 @@ public void completeVersion_Version(EObject model, Assignment assignment, Conten
+	private IResource getResource(EObject model) {
+		URI modelUri = model.eResource().getURI();
+		IResource resource=null;
+		if(modelUri.isFile()) {
+			IFile[] files = ResourcesPlugin.getWorkspace().getRoot().findFilesForLocationURI(java.net.URI.create(modelUri.toString()));
+			for (IFile iFile : files) {
+				if(iFile.exists()) {
+					resource=iFile;
+					break;
+				}
+			}
+		}else if(modelUri.isPlatform()) {
+			resource = ResourceUtils.convertEResourceToPlatformResource(model.eResource());
+		}
+		return resource;
+	}
 
-							@Override
-							public boolean apply(String extension) {
-								return filename.endsWith("." + extension);
-							}
+	private boolean isWindowsHd(String prefix) {
+		return prefix.length()==2 && prefix.charAt(1)==':' && LilyPondConstants.IS_WINDOWS;
+	}
 
-						});
-					}
+	private String getImportSearchPrefix(ContentAssistContext context) {
+		String prefix = context.getPrefix();
+		String searchPrefix=prefix;
+		if(searchPrefix.length()>0) {
+			searchPrefix=prefix.substring(1);//trim quote
+			//ensure we get the correct parent directory
+			if(searchPrefix.isEmpty() || searchPrefix.endsWith("/")) {
+				searchPrefix=searchPrefix+"X";
+			} else if(searchPrefix.endsWith("..")) {
+				searchPrefix=searchPrefix+"/X";
+			} else if(isWindowsHd(searchPrefix)) {
+				//windows hd 
+				searchPrefix=searchPrefix+"/X";
+			}
+		}
+		return searchPrefix;
+	}
 
-				});
-				for (String child : children) {
-					acceptor.accept(createCompletionProposal(QUOTE+child+QUOTE, context));
-				}
+	private List<String> getSearchPaths(IResource resource, boolean isAbsolute){
+		List<String> paths = Lists.newArrayList();
+		paths.add(resource.getParent().getLocation().toOSString());
+		if(!isAbsolute) {
+			paths.add(LilyPondImportUriResolver.getDefaultSearchUri(lilyPondPathProvider.getLilyPondPath()).getPath());
+			paths.addAll(Lists.newArrayList(lilyPondPathProvider.getSearchPaths()));
+		}
+		return paths;
+	}
+
+	private boolean propose(File f) {
+		return f.isDirectory() || fileFilter.accept(null, f.getName());
+	}
+
+	private String mergePrefixAndProposedName(String prefix, String proposedName) {
+		int slashIndex=prefix.lastIndexOf('/');
+		if(isWindowsHd(prefix) || prefix.endsWith("..")){
+			return prefix+"/"+proposedName;
+		}else if(slashIndex>=0) {
+			return prefix.substring(0, slashIndex+1)+proposedName;
+		} else {
+			return proposedName;
+		}
+	}
+
+	private void createIncludeProposal(String proposal, File file, ContentAssistContext context, ICompletionProposalAcceptor acceptor) {
+		if(proposal!=null) {
+			Image image=null;
+			ContentAssistContext contextToUse=context;
+			int priority=400;
+			if(file.isDirectory()) {
+				proposal=proposal+"/";
+				priority=300;
+				image=folderImage;
+			}else {
+				image=fileImage;
+			}
+			ICompletionProposal prop = createCompletionProposal(QUOTE+proposal+QUOTE,  new StyledString(file.getName()), image, priority, contextToUse.getPrefix(), contextToUse);
```

Comment:
I think the second quote should be omitted, since it is typically already there due to quick edit when invoking content assist. E.g:
`\include <caret>` -> type `"` -> `\include "<caret>"`

This would also make the next block (workaround for directories) unnecessary.

Reply:
I think the implementation is correct (at least more so than it would be with the modification). A proposal should insert a valid input (no matter what comes after the current cursor position). For an include a string is expected, so a complete string should be inserted. That auto edit inserts the closing quotes is a different matter (which is done so that the remainder of the document is not messed up).

With the proposed modification invoking code completion directly after `include` would yield an incomplete string. Invoking code completion in `include "ara<cared>bic.ly"` with ctrl being pressed (replacing the rest of the string) would remove the closing quote. In both cases the behaviour is incorrect.

If in java you invoke code completion in `Stri<caret>ng` another ng is inserted although it is already present. It's the same with the closing quote. If you do not want the duplicate ng/quote, you have to press ctrl while applying the proposal.

For directory navigation the current implementation works perfectly, because you can keep the ctrl key pressed the entire time (invoking code completion, applying a proposal and invoke the next completion).

Reply:
You are right, this is consistent with the other completions (e.g. version), and it can be circumvented by pressing Ctrl (Cmd).